libgnucash/engine/Account.cpp

```diff
@@ -6226,6 +6227,39 @@ gboolean xaccAccountRegister (void)
     return qof_object_register (&account_object_def);
 }
 
+
+using AccountSet = std::unordered_set<Account*>;
```

Comment:
I suppose you're using a set instead of a vector to enforce uniqueness, but what's the point of hashing the ptr? That is, why use an unordered_set instead of just a set?

Reply:
https://www.geeksforgeeks.org/unordered_set-in-cpp-stl/

Unordered set uses a hash table and set uses a red black tree, which needs rebalancing after each insertion?

Reply:
OK, but we've already demonstrated elsewhere that the C++ std library hash functions are expensive. You could fix that by providing hash and compare functions (compiles but not tested):
```
#include <unordered_set>
#include <cstdint>
/*...*/
constexpr auto hasher = [](const Account* const acct) { auto rv =std::reinterpret_cast<uintptr_t>(acct); };
constexpr auto compare = [](const Account* const lhs, const Account* const rhs) { return lhs == rhs; };
using AccountSet = std::unordered_set<Account*, hasher, compare>;
```

Reply:
not sure it's worth while the added complexity: https://stackoverflow.com/questions/20953390/what-is-the-fastest-hash-function-for-pointers

Reply:
Proves my point. What complexity are you complaining about? (BTW, I think his benchmarking is skewed by the assumption that all of the pointers are allocated from a single-type pool like GSlice so the pointers are aligned on `sizeof(T)`.  That's not the case in GnuCash so use `sizeof(unintptr_t)` to detect 32-bitness:
```
constexpr size_t shift_size = sizeof(uintptr_t) == 4 ? 1 : 3;
constexpr auto hasher = [](const Account* const acct) { auto rv = std::reinterpret_cast<uintptr_t>(acct) >> shift_size; };
```

OTOH, you're accumulating the Accounts by traversing the tree. There should be no duplicates and a std::vector is sufficient and much simpler.

Reply:
> Proves my point. What complexity are you complaining about?

I mean the complexity of offering a custom hasher.

> OTOH, you're accumulating the Accounts by traversing the tree. There should be no duplicates and a std::vector is sufficient and much simpler.

There's plenty of duplicates. Consider the acct tree

```
A
A:B
A:B:C
A:B:D
A:E
A:E:F
A:G
```

calling `gnc_accounts_and_all_descendants` with GList containing A,B,C,E,F,G will traverse all descendants of each therefore B,C,D,E,F will be processed multiple times.

The current guile code is much worse -- it calls descendants for all accounts given, each time 
generating a GList of the descendants = O(N), converting to SCM list = O(N), appends them = O(N), sorts them = O(NLogN), then traverses again eliminating duplicates = O(N). Then the gc kicks in and mops up the mess. The CPP code as above is a major speedup. To micro-optimize further seems unnecessary.

Reply:
>  it calls descendants for all accounts given

So does this

> converting to SCM list = O(N)

Saved by not using Scheme for the rest, but then spent by converting the std::set into a GList for sorting., albeit for a possibly smaller N.

>  sorts them = O(NLogN), then traverses again eliminating duplicates = O(N)

The sort set is still there, though also for a smaller N. The algorithm isn't that different, really. The only supposed savings is the remove duplicates step, but absent any profiling there's no reason to assume that it's more expensive than constructing the std::unordered_set.

> Then the gc kicks in and mops up the mess.

GC, like `free`, is cheap. It's the allocation that's expensive and they're both allocating. Plus GC runs asynchronously in its own thread, possibly on a different core, while the std::unordered_set element destructors and frees run synchronously in this thread.

> calling gnc_accounts_and_all_descendants with GList containing A,B,C,E,F,G will traverse all descendants of each therefore B,C,D,E,F will be processed multiple times. The output will be A,B,C,D,E,F,G.

That's insane, but I guess it's the result of selecting A and all of its children in the account selector and then mindlessly running `gnc_account_foreach_descendant` on the result.

It's a poorly designed brute-force algorithm. A little analysis up front would reduce N and is conceptually more likely to save time: Walk the tree backwards, eliminating any child accounts: In your example, C's and D's parent is B, so no need to traverse those two. F's parent is E, ditto. The remaining B, E, and G all have parent A, so they're gone too. We just need to call gnc_account_for_each_descendant on A and we know there will be no duplicates so we can put the results directly in a GList and then sort it.

But absent actual profiling data maybe this whole exercise is bike-shedding. Is gnc:accounts_and_all_descendants on the hot trace for any of the reports that use it? Does this conversion to C++ produce a significant savings in running those reports?

Reply:
> But absent actual profiling data maybe this whole exercise is bike-shedding. Is gnc:accounts_and_all_descendants on the hot trace for any of the reports that use it? Does this conversion to C++ produce a significant savings in running those reports?

It's not a hot function, however it's an embarrassing function in scheme. Many speedups have taken place in the past 3 years (aka death by thousand cuts) so this one is a candidate too. The predecessor https://github.com/Gnucash/gnucash/blob/868489b1c4f9b1f64fe61130e6ae28b4bad16f22/src/report/report-system/report-utilities.scm#L148 was even worse, by keeping duplicates. The resulting expanded accountlist was used for QofQuery.

Reply:
I think there's no need to utf8 sort this account list which generally is fed into QofQiery. This will effect a small change in tests.

Reply:
Definitely. In half the calls the list is passed to `gnc:accounts-get-commodities` so it definitely doesn't need to be sorted. Then there's budget.scm and cash-flow.scm where the result is sorted again with `(sort accounts gnc:account-full-name<?))`.  It seems to get fed to a query only in examples/average-balance.scm and examples/daily-reports. scm. In category-barcharts.scm there's a call where its rv is passed to a lambda that creates an alist with the account as the key, you might want to sort that.

Reply:
Ok I'll remove the expensive utf8 sorting. This is the only functional difference with the scheme version, and causes a small change in tests. If any reports eventually show accounts in a wrong order it'll be easy to reinstate sorting within the report itself.